### ISSUE TITLE
parsl/app: ImportError: cannot import name 'RemoteException'

### ISSUE THREAD

**Issue by ahayschi** [Parsl/parsl#228]
Context: running a simple hello world python app, ipp executor, and ec2 provider.

Parsl log was showing 
```ipyparallel.error.RemoteError: ImportError(cannot import name 'RemoteException')```

and ipengine log was showing
```
2018-04-25 03:58:34.776 [IPEngineApp] apply_request: {'parent_header': {}, 'content': {}, 'msg_id': 'bbc8c92c-c590d9f1b910ff0f72796a30', 'header': {'msg_id': 'bbc8c92c-c590d9f1b910ff0f72796a30', 'session': '6fc4c785-746b4d95fd0735d2eb9428b7', 'version': '5.3', 'date': datetime.datetime(2018, 4, 25, 3, 56, 51, 55382, tzinfo=tzlocal()), 'msg_type': 'apply_request', 'username': 'ahays'}, 'buffers': [<memory at 0x7f5ef0042348>, <memory at 0x7f5ef0042408>], 'msg_type': 'apply_request', 'metadata': {'targets': [], 'retries': 0, 'follow': [], 'timeout': 0.0, 'after': []}}
2018-04-25 03:58:34.965 [IPEngineApp] Exception in apply request:
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<string> in <module>()

/Users/ahays/code/parsl/parsl/app/errors.py in wrapper(*args, **kwargs)

ImportError: cannot import name 'RemoteException'
```
Removing `from parsl.app.errors import RemoteException` in the wrapper fixed the issue.

**Comment by annawoodard** [comment]
Thanks for the report, @ahayschi! I will look at this today.

**Comment by yadudoc** [comment]
@ahayschi Can you please add the runinfo/RUNDIR/parsl.log ?

**Comment by ahayschi** [comment]
Sure

```
2018-04-24 23:56:48 parsl.dataflow.dflow:75 [INFO]  Parsl version: 0.5.0
2018-04-24 23:56:48 parsl.dataflow.dflow:76 [INFO]  Libsubmit version: 0.3.2
2018-04-24 23:56:48 parsl.dataflow.config_defaults:19 [DEBUG]  { 'controller': {'mode': 'auto', 'publicIp': 'xxx', 'reuse': False},
  'globals': { 'appCache': True,
               'checkpoint': None,
               'checkpointFiles': None,
               'checkpointPeriod': None,
               'lazyErrors': True,
               'strategy': 'simple',
               'usageTracking': True},
  'sites': [ { 'auth': {'channel': None},
               'execution': { 'block': { 'initBlocks': 1,
                                         'maxBlocks': 10,
                                         'minBlocks': 0,
                                         'nodes': 1,
                                         'options': { 'imageId': 'ami-916f59f4',
                                                      'keyName': 'dev-us-east-2',
                                                      'stateFile': 'awsproviderstate.json'},
                                         'parallelism': 0.75,
                                         'taskBlocks': 1,
                                         'walltime': '00:20:00'},
                              'container': {'image': None, 'type': None},
                              'executor': 'ipp',
                              'provider': 'aws',
                              'scriptDir': '/Users/ahays/code/parsl/runinfo/057'},
               'site': 'aws-east'}]}
2018-04-24 23:56:48 parsl.dataflow.usage_tracking.usage:58 [DEBUG]  Tracking status: True
2018-04-24 23:56:48 parsl.dataflow.usage_tracking.usage:59 [DEBUG]  Testing mode   : False
2018-04-24 23:56:48 parsl.dataflow.memoization:60 [INFO]  App caching initialized
2018-04-24 23:56:48 parsl.execution_provider.provider_factory:76 [DEBUG]  Constructing site : aws-east 
2018-04-24 23:56:48 parsl.execution_provider.provider_factory:91 [DEBUG]  Created channel : None
2018-04-24 23:56:49 parsl.execution_provider.provider_factory:104 [DEBUG]  Created execution_provider : <EC2 Execution Provider for site:aws-east>
2018-04-24 23:56:49 parsl.execution_provider.provider_factory:114 [DEBUG]  Starting controller
2018-04-24 23:56:49 parsl.executors.ipp_controller:40 [DEBUG]  Starting ipcontroller, baseDir:/Users/ahays/code/parsl/runinfo/057
2018-04-24 23:56:49 parsl.executors.ipp_controller:100 [DEBUG]  Start opts: ['ipcontroller', '--ipython-dir=/Users/ahays/code/parsl/runinfo/057', '--ip=*', '--profile=aws-east', '', '', '--location=xxx']
2018-04-24 23:56:49 parsl.execution_provider.provider_factory:122 [DEBUG]  Controller engine file : /Users/ahays/code/parsl/runinfo/057/profile_aws-east/security/ipcontroller-engine.json
2018-04-24 23:56:49 parsl.execution_provider.provider_factory:123 [DEBUG]  Controller client file : /Users/ahays/code/parsl/runinfo/057/profile_aws-east/security/ipcontroller-client.json
2018-04-24 23:56:49 parsl.executors.ipp:135 [DEBUG]  Waiting for /Users/ahays/code/parsl/runinfo/057/profile_aws-east/security/ipcontroller-client.json
2018-04-24 23:56:49 parsl.executors.ipp:177 [DEBUG]  Starting IpyParallelExecutor with provider:<EC2 Execution Provider for site:aws-east>
2018-04-24 23:56:51 parsl.executors.ipp:181 [DEBUG]  Launched block : 0:i-062f8e69e0d05d6f5
2018-04-24 23:56:51 parsl.executors.ipp:196 [DEBUG]  Starting executor
2018-04-24 23:56:51 parsl.execution_provider.provider_factory:135 [DEBUG]  Created executor : <IPP Executor for site:aws-east>
2018-04-24 23:56:51 parsl.dataflow.strategy:119 [DEBUG]  Scaling strategy: simple
2018-04-24 23:56:51 parsl.dataflow.dflow:134 [DEBUG]  Using executors: {'aws-east': <IPP Executor for site:aws-east>}
2018-04-24 23:56:51 parsl.dataflow.dflow:483 [INFO]  Task 0 submitted for App hello_world, waiting on tasks []
2018-04-24 23:56:51 parsl.executors.ipp:211 [DEBUG]  Got args : (<function hello_world at 0x10cc16f28>,),
2018-04-24 23:56:51 parsl.executors.ipp:212 [DEBUG]  Got kwargs : {},
2018-04-24 23:56:51 parsl.dataflow.dflow:323 [INFO]  Task 0 launched on site aws-east
2018-04-24 23:56:51 parsl.dataflow.dflow:506 [DEBUG]  Task 0 launched with AppFut:<AppFuture at 0x10f95e048 state=pending>
2018-04-24 23:56:52 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:56:52 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:56:57 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:56:57 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:02 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:02 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:07 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:07 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:12 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:12 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:17 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:17 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:22 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:22 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:27 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:27 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:32 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:32 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:37 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:37 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:42 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:42 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:47 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:47 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:52 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:52 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:57:57 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:57:57 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:02 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:58:02 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:07 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:58:07 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:12 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:58:12 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:17 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:58:17 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:22 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:58:22 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:27 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:58:27 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:32 parsl.dataflow.strategy:178 [DEBUG]  Min:0 initBlocks:1 Max:10
2018-04-24 23:58:32 parsl.dataflow.strategy:182 [DEBUG]  Tasks:1 Slots:1 Parallelism:0.75
2018-04-24 23:58:34 parsl.dataflow.dflow:186 [ERROR]  Task 0 failed
Traceback (most recent call last):
  File "/Users/ahays/code/parsl/parsl/dataflow/dflow.py", line 183, in handle_update
    future.result()
  File "/Users/ahays/.pyenv/versions/3.5.2/lib/python3.5/concurrent/futures/_base.py", line 398, in result
    return self.__get_result()
  File "/Users/ahays/.pyenv/versions/3.5.2/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/Users/ahays/.pyenv/versions/3.5.2/lib/python3.5/site-packages/ipyparallel/client/asyncresult.py", line 226, in _resolve_result
    raise r
ipyparallel.error.RemoteError: ImportError(cannot import name 'RemoteException')
2018-04-24 23:58:34 parsl.dataflow.dflow:203 [INFO]  Task 0 failed after 0 retry attempts
2018-04-24 23:58:34 parsl.dataflow.dflow:544 [INFO]  DFK cleanup initiated
2018-04-24 23:58:34 parsl.dataflow.dflow:562 [INFO]  Terminating flow_control and strategy threads
```

**Comment by yadudoc** [comment]
@ahayschi, I'm able to reproduce this error. Here's what is happening:

You are running a fairly up to date version of Parsl from github. On EC2 remote side, you have got parsl v0.5.0 installed from PyPI. The recently added wrapper code imports a module that is not available on the remote side which uses v0.5.0. This code path is hit only when the app code raises an error. So the problem simply boils down a version mismatch.

We'll put out a point release soon (targetting next week) which will fix this issue.

**Comment by ahayschi** [comment]
Anna and Yadu thanks for looking into this and reporting back. This seems like a silly mistake on my part but something I will keep in mind.

Would a version checked performed when an engine connects be useful? Would allow the workflow to fail early and return an explicit error. This would also be useful for mismatched Python versions too I imagine.
